File: lib/models/group.dart
class Group {
  String id;
  String name;
  int leaderId;
  String leaderName;
  int members;
  String groupNumber;
  String desc;
  int status;
  String created;
  String updated;

  Group({
    this.id = '',
    this.name = '',
    this.leaderId = 0,
    this.leaderName = '',
    this.members = 0,
    this.groupNumber = '',
    this.desc = '',
    this.status = 0,
    this.created = '',
    this.updated = ''
  });

  void clone(Group group) {
    id = group.id;
    name = group.name;
    leaderId = group.leaderId;
    leaderName = group.leaderName;
    members = group.members;
    groupNumber = group.groupNumber;
    desc = group.desc;
    status = group.status;
    created = group.created;
    updated = group.updated;
  }

  factory Group.fromJson(Map<String, dynamic> jsonData) {
    return Group(
      id: jsonData['id'] ?? '',
      name: jsonData['name'] ?? '',
      leaderId: jsonData['leader_id'] ?? 0,
      leaderName: jsonData['leaderName'] ?? '',
      members: jsonData['members'] ?? 0,
      groupNumber: jsonData['group_number'] ?? 0,
      desc: jsonData['desc'] ?? '',
      status: jsonData['status'] ?? 0,
      created: jsonData['created'] ?? '',
      updated: jsonData['updated'] ?? ''
    );
  }
}

class GroupList {
  GroupList({
    this.currentPage = 0,
    this.firstPageUrl = '',
    this.from = 0,
    this.lastPage = 0,
    this.lastPageUrl = '',
    this.nextPageUrl = '',
    this.path = '',
    this.perPage = '',
    this.prevPageUrl = '',
    this.to = 0,
    this.total = 0,
    this.data
  });

  int currentPage;
  String firstPageUrl;
  int from;
  int lastPage;
  String lastPageUrl;
  String nextPageUrl;
  String path;
  String perPage;
  String prevPageUrl;
  int to;
  int total;
  List<Group>? data;

  factory GroupList.fromJson(Map<String, dynamic> jsonData) {
    List<Group> data = [];
    if (jsonData['data'] != null) {
      data = (jsonData['data'] as List<dynamic>)
          .map((e) => Group.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return GroupList(
      currentPage: jsonData['current_page'] ?? 1,
      firstPageUrl: jsonData['first_page_url'] ?? '',
      from: jsonData['from'] ?? 1,
      lastPage: jsonData['last_page'] ?? 1,
      lastPageUrl: jsonData['last_page_url'] ?? '',
      nextPageUrl: jsonData['next_page_url'] ?? '',
      path: jsonData['path'] ?? '',
      perPage: jsonData['per_page'] ?? '0',
      prevPageUrl: jsonData['prev_page_url'] ?? '',
      to: jsonData['to'] ?? 1,
      total: jsonData['total'] ?? 0,
      data: data,
    );
  }
}

File: application/controllers/admin/Group.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

class Group extends BaseAdminController {
	
	// constructor
	function __construct() {
		parent::__construct();
		
		$this->load->model('group_model');
		$this->load->model('user_model');
	}

    public function index()
    {
        $page = ($this->uri->segment(4)) ? $this->uri->segment(4) : 1;
        $search = $this->input->get('search');
        $limit = $this->input->get('limit');
        $leaderId = $this->input->get('leader_id');
        $status = $this->input->get('status');
        $parameter['page'] =  $page;
        $parameter['search'] =  $search;
        $parameter['limit'] =  $limit;
        $parameter['leader_id'] =  $leaderId;
        $parameter['status'] =  $status;
        $mData = $this->group_model->list($parameter);
        $data['title'] = "OSCARS BALLOT- Groups";
        $data['isLogin'] = isset($this->session->userdata['sign']);
        $data['isActiveGroupPage'] = true;
        $data['list'] = $mData['result'];
        $pagination = $this->pagination($mData['total'], $mData['limit'], $mData['page'], '/admin/group/index');
        $data['pagination'] = $pagination['pagination'];
        $data['pagination_string'] = $pagination['pagination_string'];
        $data['page_sizes'] = config_item('page_sizes');
        $data['users'] = $this->user_model->getAllUsers();
        $this->load->view('admin/group/list', $data);
    }

    public function add()
    {
        $data['title'] = "OSCARS BALLOT- GROUP ADD";
        $data['isLogin'] = isset($this->session->userdata['sign']);
        $data['isActiveGroupPage'] = true;
        $data['edit'] = false;
        $data['users'] = $this->user_model->getAllUsers();
        $this->load->view('admin/group/add_edit', $data);
    }

    public function store()
    {
        $user = OscarAuth::getInstance()->user();
        $data['name'] = $this->input->post('name');
        $data['desc'] = $this->input->post('desc');
        $data['status'] = $this->input->post('status');
        $data['leader_id'] = $this->input->post('leader_id');
        $data['user_id'] = $user['id'];
        $data['group_number'] = $this->input->post('group_number');
        $data['group_password'] = $this->input->post('group_password');
        $data['created'] = date('y-m-d H:i:s');
        $data['updated'] = date('y-m-d H:i:s');
        $this->group_model->add($data);

        return redirect('admin/group/index');
    }

    public function edit()
    {
        $data['title'] = "OSCARS BALLOT- GROUP EDIT";
        $data['isLogin'] = isset($this->session->userdata['sign']);
        $data['isActiveGroupPage'] = true;
        $data['edit'] = true;
        $id = $this->uri->segment(4);
        $group = $this->group_model->find($id);
        $data['group'] = $group;
        $data['users'] = $this->user_model->getAllUsers();
        $this->load->view('admin/group/add_edit', $data);
    }

    public function update()
    {
        $id = $this->input->post('id');
        $data['name'] = $this->input->post('name');
        $data['desc'] = $this->input->post('desc');
        $data['status'] = $this->input->post('status');
        $data['leader_id'] = $this->input->post('leader_id');
        $data['group_number'] = $this->input->post('group_number');
        $data['group_password'] = $this->input->post('group_password');
        $data['updated'] = date('y-m-d H:i:s');
        $this->group_model->update($id, $data);

        return redirect('admin/group/index');
    }


    public function delete()
    {
        $id = $this->input->post('id');
        $this->group_model->delete($id);
        return json_encode([
            'success' => true
        ]);
    }

    public function bulkDelete()
    {
        $ids = explode(',', $this->input->post('ids'));
        if (!empty($ids)) {
            foreach($ids as $id) {
                $this->group_model->delete($id);
            }
        }

        return json_encode([
            'success' => true
        ]);
    }

    public function members()
    {
        $groupId = $this->input->get('group_id');
        $search = $this->input->get('search');
        $limit = $this->input->get('limit');
        $status = $this->input->get('status');
        $parameter['search'] =  $search;
        $parameter['limit'] =  $limit;
        $parameter['status'] =  $status;
        $mData = $this->group_model->members($groupId, $parameter);
        $data['members'] = $mData['result'];
        $data['total'] = $mData['total'];
        $pagination = $this->pagination($mData['total'], $mData['limit'], $mData['page'], '/admin/group/members');
        $data['pagination'] = $pagination['pagination'];
        $data['pagination_string'] = $pagination['pagination_string'];
        $data['page_sizes'] = config_item('page_sizes');
        $this->load->view('admin/group/members', $data);
    }

    public function addMember()
    {
        $data['users'] = $this->user_model->getAllUsers();
        $this->load->view('admin/group/add_member', $data);
    }


    public function storeMember()
    {
        $memberId = $this->input->post('member_id');
        $groupId = $this->input->post('group_id');
        $status = Group_model::STATUS_GROUP_PENDING;
        $groupMember = $this->group_model->addMember($groupId, $memberId, $status);
        return json_encode([
            'success' => true,
            'data' => $groupMember
        ]);
    }

    public function deleteMember()
    {
        $memberId = $this->input->post('member_id');
        $this->group_model->deleteMember($memberId);
        return json_encode([
            'success' => true
        ]);
    }

    public function bulkDeleteMembers()
    {
        $ids = explode(',', $this->input->post('member_ids'));
        $this->group_model->removeMembers($ids);
        return json_encode([
            'success' => true
        ]);
    }
}